src/Confluent.Kafka/Config.cs

```diff
@@ -0,0 +1,903 @@
+        Fifo,
+
+        /// <summary>
+        ///     Lifo
+        /// </summary>
+        Lifo
+    }
+
+    /// <summary>
+    ///     Partitioner enum values
+    /// </summary>
+    public enum PartitionerType
+    {
+        /// <summary>
+        ///     Random
+        /// </summary>
+        Random,
+
+        /// <summary>
+        ///     Consistent
+        /// </summary>
+        Consistent,
+
+        /// <summary>
+        ///     Consistent_Random
+        /// </summary>
+        Consistent_Random,
+
+        /// <summary>
+        ///     Murmur2
+        /// </summary>
+        Murmur2,
+
+        /// <summary>
+        ///     Murmur2_Random
+        /// </summary>
+        Murmur2_Random
+    }
+
+    /// <summary>
+    ///     AutoOffsetReset enum values
+    /// </summary>
+    public enum AutoOffsetResetType
+    {
+        /// <summary>
+        ///     Latest
+        /// </summary>
+        Latest,
+
+        /// <summary>
+        ///     Earliest
+        /// </summary>
+        Earliest,
+
+        /// <summary>
+        ///     Error
+        /// </summary>
+        Error
+    }
+
+    /// <summary>
+    ///     SaslMechanism enum values
+    /// </summary>
+    public enum SaslMechanismType
+    {
+        /// <summary>
+        ///     GSSAPI
+        /// </summary>
+        Gssapi,
+
+        /// <summary>
+        ///     PLAIN
+        /// </summary>
+        Plain,
+
+        /// <summary>
+        ///     SCRAM-SHA-256
+        /// </summary>
+        ScramSha256,
+
+        /// <summary>
+        ///     SCRAM-SHA-512
+        /// </summary>
+        ScramSha512
+    }
+
+    /// <summary>
+    ///     Configuration common to all clients
+    /// </summary>
+    public class ClientConfig : IEnumerable<KeyValuePair<string, string>>
+    {
+        /// <summary>
+        ///     Initialize a new empty <see cref="ClientConfig" /> instance.
+        /// </summary>
+        public ClientConfig() {}
+
+        /// <summary>
+        ///     Initialize a new <see cref="ClientConfig" /> instance based on
+        ///     an existing <see cref="ClientConfig" /> instance.
+        /// </summary>
+        public ClientConfig(ClientConfig config) { this.properties = new Dictionary<string, string>(config.ToDictionary(a => a.Key, a => a.Value)); }
+
+        /// <summary>
+        ///     Initialize a new <see cref="ClientConfig" /> instance based on
+        ///     an existing key/value pair collection.
+        /// </summary>
+        public ClientConfig(IEnumerable<KeyValuePair<string, string>> config) { this.properties = new Dictionary<string, string>(config.ToDictionary(a => a.Key, a => a.Value)); }
+
+        /// <summary>
+        ///     Set a configuration property using a string key / value pair.
+        /// </summary>
+        /// <remarks>
+        ///     Two scenarios where this is useful: 1. For setting librdkafka
+        ///     plugin config properties. 2. You are using a different version of 
```

Comment:
Make sure this dict retains ordering

Reply:
A dictionary - by definition - does not retain ordering. A configuration interface that relies on ordering is unusual and necessarily more difficult to use / less terse. we should discuss this.

Reply:
I agree, but `plugin.library.paths` is an exception that we'll need to support.
The Python and Go clients extract and apply plugin.library.paths before any other configuration is set, which is probably a good workaround. (cc @rnpridgeon )
https://github.com/confluentinc/confluent-kafka-python/blob/master/confluent_kafka/src/confluent_kafka.c#L1582